### ISSUE TITLE
Media3 MediaSession custom error messages

### ISSUE THREAD

**Issue by abouda** [androidx/media#543]
Is there an equivalent in media3 for these?
- `MediaSessionConnector.setCustomErrorMessage()` to send a custom error message regardless of player state.
- `MediaSessionConnector.setErrorMessageProvider(PlayerErrorMessageProvider)` to convert PlaybackException to a custom error message for connected controllers.

**Comment by marcbaechinger** [comment]
There is not such functionality in Media3. 

Who would be the consumer of the error messages?

**Comment by abouda** [comment]
- The consumer for `MediaSessionConnector.setCustomErrorMessage()` is AndroidAuto. It sends respective error messages to be displayed on the AndroidAuto screen when something goes wrong with a play action. (for example user not logged in, no book available for resumption, playback error custom message). Right now with media3, the error message just says "Source error" and I want it to be more user friendly.

- The consumer for `MediaSessionConnector.setErrorMessageProvider(PlayerErrorMessageProvider)` is the MediaControllerCompat in my app (I haven´t migrated that part to media3 yet). I map the PlaybackException to a respective error message that is user friendly, and display it in an alert dialog.
I see that media3 MediaController has a onPlayerError(PlaybackException) so I am guessing that could be the solution and handle the PlaybackException on the MediaBrowser side.

**Comment by abouda** [comment]
Also another point is although I could handle the error `onPlayerError(PlaybackException)` in my app, it does not solve it for legacy controllers like AndroidAuto.

**Comment by marcbaechinger** [comment]
Yes, this is a regression compared with what's possible with the legacy API. 
We need to look into this. Marked as enhancement.

**Comment by Bwaim** [comment]
@marcbaechinger do you hav any news on this ?

**Comment by marcbaechinger** [comment]
Yeah, sorry no news yet.

I don't think this will be part of 1.3 I'm afraid. We will update this issue as soon as we pushed a change that adds this.

**Comment by Flyktsodan** [comment]
Our app update got rejected in playstore review because we didn't send error messages when using android auto. In other words this should be a blocker for many media app(s).

**Comment by benjaminVadon** [comment]
@marcbaechinger  have you any update about this limitation ? Is it planned for the next version ?

**Comment by marcbaechinger** [comment]
This is planned for 1.4 which is the next release.

**Comment by marcbaechinger** [comment]
The commit above added an API to send non-fatal errors to a `MediaController`. 

When sent to all controllers or specifically to the media notification controller (`MediaSession.getMediaNotificationControllerInfo()`), the error code, message and extras are included in the `PlaybackStateCompat` with `state=STATE_ERROR` (7) and sent to the platform/legacy session from where a platform or legacy controller (for instance Android Auto) can read it.

**Comment by icutvaric** [comment]
Hello!

Any updates on this? 

Or any suggestion how to do authentication on Automotive OS with Media3: https://developer.android.com/training/cars/media/automotive-os#require-sign-in?

Thanks!

**Comment by kelmer44** [comment]
Is there a reason why a resource Id is necessary? Could the method not accept just a plain string as well?

**Comment by marcbaechinger** [comment]
This API has changed. The change that includes this is on its way into the main branch.

There will be a `SessionError` instead of the three parameters: `MediaSession.sendError(ControllerInfo, SessionError)`. The message argument of the constructor of `MediaSession` is a String instead of a resource ID. Please use a localized string either way for the benefit of your users.

**Comment by kelmer44** [comment]
@marcbaechinger thats great! I do plan on using a localized string, only we dont use the android resources in our project, but an alternative solution

**Comment by chillbrodev** [comment]
> This is planned for 1.4 which is the next release.

Is there a potential release date for 1.4?

**Comment by marcbaechinger** [comment]
We are currently preparing the beta01. The stable release of 1.4.0  is expected to be in mid July.

**Comment by Flyktsodan** [comment]
I tried the latest version and it works in terms of showing a error message. The android auto companion app still shows "Getting your selection" and users have to navigate back and try again. Is there any way of communicating that the ListenableFuture<MediaItemsWithStartPosition> failed and it should close the "player view" in android auto? See screenshot for reference: 
<img width="473" alt="Screenshot 2024-08-21 at 11 17 50" src="https://github.com/user-attachments/assets/e5f7f196-741a-4030-8536-d35fdefda010">

**Comment by marcbaechinger** [comment]
Thanks for reporting.

What did you do this in the old world to make 'close the "player view" in android auto'?

Not sure if this is what you are asking about, but I mean to remember that you can send a `PendingIntent` back to Auto that then opens that activity for resolving the error in some ways. Is this what you are asking for? 

This is documented here: https://developers.google.com/cars/design/create-apps/media-apps/signin-flow

If this is the case, then you can add the same extras key/value to the `LibraryResult` that does report the authentication error code. Something like this:

```
return Futures.immediateFuture(
  LibraryResult.ofError(
    LibraryResult.RESULT_ERROR_SESSION_AUTHENTICATION_EXPIRED,
    LibraryParams.Builder()
       .setExtras(getExpiredAuthenticationResolutionExtras()).build()
  )
)

private fun getExpiredAuthenticationResolutionExtras(): Bundle {
        return Bundle().also {
            it.putString(
                EXTRAS_KEY_ERROR_RESOLUTION_ACTION_LABEL_COMPAT,
                getString(R.string.login_button_label))
            val signInIntent = Intent(this, SignInActivity::class.java)
            @OptIn(UnstableApi::class)
            val flags = if (Util.SDK_INT >= 23) PendingIntent.FLAG_IMMUTABLE else 0
            it.putParcelable(
                EXTRAS_KEY_ERROR_RESOLUTION_ACTION_INTENT_COMPAT,
                PendingIntent.getActivity(this, /* requestCode= */ 0, signInIntent, flags))
        }
    }
```

**Comment by Flyktsodan** [comment]
Thanks for the quick replay, with our old ExoPlayer implementation we would authenticate for playback, if it failed we would call mediaSessionConnector.setCustomErrorMessage(errorMessage). That resulted in a message being shown in the android auto ui (see attached screenshot). 

With media3, we authenticate for playback in onSetMediaItems (when it's called from an Android auto controller). The difference here is calling mediaSession.sendError() displays a toast instead and the screen remains in "Getting your selecting" (as shown in first screenshot). I've tried returning a Futures.immediateFailedFuture in onSetMediaItems but it doesn't have any effect. The only way I can get the screen to show the "Source error" is by returning a broken MediaItem that can't be played.

<img width="581" alt="Screenshot 2024-08-21 at 12 06 36" src="https://github.com/user-attachments/assets/4a313cf8-7f4e-43d6-9f0e-d852681edc36">

**Comment by Flyktsodan** [comment]
@marcbaechinger can you think of any way in how to replicate the old behaviour described above? 
I'm worried app will be rejected in playstore review if this is not fixed somehow.

**Comment by marcbaechinger** [comment]
I think Media3 is currently missing an API to send a custom fatal exception in an easy way. 

The current version allows to send non-fatal errors as you do and it allows to customize error code and error message of a `PlaybackException`. However, there isn't an easy way do originate a fatal `PlaybackException` that isn't coming from the `Player`.

https://developer.android.com/media/media3/session/control-playback#error-handling

Sorry for not having a better answer than that we should prioritize this to provide this with the next release.

**Comment by manthan34** [comment]
> Thanks for the quick reply, with our old ExoPlayer implementation we would authenticate for playback, if it failed we would call mediaSessionConnector.setCustomErrorMessage(errorMessage). That resulted in a message being shown in the android auto ui (see attached screenshot).
> 
> With media3, we authenticate for playback in onSetMediaItems (when it's called from an Android auto controller). The difference here is calling mediaSession.sendError() displays a toast instead and the screen remains in "Getting your selection" (as shown in first screenshot). I've tried returning a Futures.immediateFailedFuture in onSetMediaItems but it doesn't have any effect. The only way I can get the screen to show the "Source error" is by returning a broken MediaItem that can't be played.
> 
> <img alt="Screenshot 2024-08-21 at 12 06 36" width="581" src="https://private-user-images.githubusercontent.com/42470133/359852579-4a313cf8-7f4e-43d6-9f0e-d852681edc36.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwODY4NjQsIm5iZiI6MTczMjA4NjU2NCwicGF0aCI6Ii80MjQ3MDEzMy8zNTk4NTI1NzktNGEzMTNjZjgtN2Y0ZS00M2Q2LTlmMGUtZDg1MjY4MWVkYzM2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTIwVDA3MDkyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRiYTg2NTI3M2Q5MDI3ZmQ4OTY3NmRlMWVhNGEwZTg1MzRjOGE5N2E3YjBmMDNjMjBkM2YwN2VkYmE1OTdiZTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.lLZUjLmSVz7WOS7Anp4IRJBLSBFPvPES-vY1L6znk24">

Hi @Flyktsodan ,

I’m facing a similar issue as mentioned in your post and would greatly appreciate your help with the following:

1. How did you manage to display a custom error message on the Android Auto screen?
2. You mentioned returning a broken `MediaItem`. Could you please provide a sample code or more details on how you handled this?

Thanks in advance!

**Comment by marcbaechinger** [comment]
https://developer.android.com/media/media3/session/control-playback#fatal-errors doesn't work in this case?

**Comment by marcbaechinger** [comment]
Thanks I have nothing to add I'm afraid. 

When you return a `LibraryResult.RESULT_ERROR_SESSION_AUTHENTICATION_EXPIRED` for one of the session callback, a fatal error is written into the `PlaybackState` of the platform session. If this isn't working please file a bug. If this isn't a solution for your use case the I'm afraid, but I can't drop everything else to work on this now.

**Comment by marcbaechinger** [comment]
Yeah, I meant returning that from `onGetChildren()` from the `MediaLibrarySession.Callback`.

I think the remaining piece of this issue is adding ways for apps to send generate their own `PlaybackException` even if there isn't a `PlaybackException` coming from the player. This is a but a remaining of the legacy library that leaks into the new API with this. In media1 policy errors in the `MediaBrowserServiceCompat` have been reported as fatal exception in the playback state. This wasn't nicely separated in concerns IMO. So we added the workaround that returning an authentication error from the `MediaLibraryCallback.onXyz()` triggers a fatal exception in the platform playback state.

So I think regarding the custom `PlaybackException` there isn't a need to create a new issue.

I'll do my best to deliver this asap.

**Comment by c-erdos** [comment]
Hi - I'm a PM on the Atlas team checking in on status of this fix? We're building a new instance of Android Auto and need to implement different error messages for different scenarios: https://www.figma.com/design/aFsWpw4nybSQQzUqA4C5fz/CarPlay-%26-Android-Auto%3A-Eng-Spec?node-id=10575-22035&t=kH0E3tldfKnDaKxU-4

I was told this is not currently possible on Android Auto. What's timeline look like for fixing?

**Comment by marcbaechinger** [comment]
https://github.com/androidx/media/commit/5d5a430756cdd42e5ae80727b36b8a8fe151085c
https://github.com/androidx/media/commit/317b40f26961c6c4db630eea618c612ac038d260


We've added `MediaSession.setPlaybackException(ControllerInfo, PlaybackException)` that allows setting custom `PlaybackException` for selected or all controllers. Please see the [JavaDoc in the main](https://github.com/androidx/media/blob/main/libraries/session/src/main/java/androidx/media3/session/MediaSession.java#L1123-L1140) branch for now to learnhow to set a fatal error.


I'm closing this issue. A Media3 app can now 1) send non-fatal policy errors that are not affecting playback, 2) rewrite and change error codes and message of actual `PlaybackException` instances thrown by the player and 3) set custom `PlaybackException` for a certain Media3 controller (use the media notification controller to set an error for the legacy session). 

We will add documentation for 3) to the official [documentation on this page](https://developer.android.com/media/media3/session/control-playback#error-handling), once we have released that change. 

The change will be included in the next release  for which the first alpha has been cut already.

If you find an problem with this or have some further input, please file another issue and describe the issue as usual as per issue template.
